File: apps/api/internal/domain/entities/api_key.go
package entities

import "time"

// MCPAPIScope grants a machine principal access to a State MCP tool category.
type MCPAPIScope string

const (
	MCPAPIScopeStateRead        MCPAPIScope = "state:read"
	MCPAPIScopeStateWrite       MCPAPIScope = "state:write"
	MCPAPIScopeCapabilityInvoke MCPAPIScope = "capability:invoke"
)

// ValidMCPAPIScopes is the complete allowlist for machine API key scopes.
var ValidMCPAPIScopes = map[MCPAPIScope]struct{}{
	MCPAPIScopeStateRead:        {},
	MCPAPIScopeStateWrite:       {},
	MCPAPIScopeCapabilityInvoke: {},
}

// APIKey is the non-secret, persisted representation of a machine credential.
// KeyVerifier is retained for authentication but is never exposed through DTOs.
type APIKey struct {
	ID               string
	TenantID         string
	Name             string
	Prefix           string
	KeyVerifier      []byte
	ProjectIDs       []string
	DefaultProjectID *string
	Scopes           []MCPAPIScope
	ExpiresAt        *time.Time
	RevokedAt        *time.Time
	LastUsedAt       *time.Time
	CreatedBy        string
	CreatedAt        time.Time
	UpdatedAt        time.Time
}

// APIKeyPrincipal is the immutable authenticated identity available to State
// MCP handlers. It deliberately excludes the verifier and raw key secret.
type APIKeyPrincipal struct {
	KeyID            string
	TenantID         string
	KeyPrefix        string
	ProjectIDs       []string
	DefaultProjectID *string
	Scopes           []MCPAPIScope
}

// HasScope reports whether the principal is allowed to use a scope.
func (p APIKeyPrincipal) HasScope(scope MCPAPIScope) bool {
	for _, granted := range p.Scopes {
		if granted == scope {
			return true
		}
	}
	return false
}

// AllowsProject reports whether a project belongs to this machine principal's
// explicit allowlist.
func (p APIKeyPrincipal) AllowsProject(projectID string) bool {
	for _, allowed := range p.ProjectIDs {
		if allowed == projectID {
			return true
		}
	}
	return false
}

File: packages/schemas/api-key.ts
import { z } from "zod"

export const mcpAPIKeyScopes = [
  "state:read",
  "state:write",
  "capability:invoke",
] as const

export const mcpAPIKeyScopeLabels = [
  { label: "Read state", value: "state:read" },
  { label: "Write state", value: "state:write" },
  { label: "Invoke capabilities", value: "capability:invoke" },
] as const

export const getMCPAPIKeyScopeLabel = (
  value: (typeof mcpAPIKeyScopes)[number],
) => mcpAPIKeyScopeLabels.find((item) => item.value === value)?.label ?? value

export const createStateMCPAPIKeySchema = z.object({
  name: z.string().trim().min(1, "API key name is required").max(255),
  projectIds: z
    .array(z.string().trim().min(1, "Project ID is required"))
    .min(1, "At least one project ID is required"),
  defaultProjectId: z.string().trim().min(1).optional(),
  scopes: z.array(z.enum(mcpAPIKeyScopes)).min(1, "Select at least one scope"),
  expiresAt: z.string().datetime({ offset: true }).optional(),
})

export type CreateStateMCPAPIKeySchemaProps = z.infer<
  typeof createStateMCPAPIKeySchema
>

const nullableDate = z.string().nullable()

export const stateMCPAPIKeyResponseSchema = z.object({
  id: z.string(),
  tenantId: z.string(),
  name: z.string(),
  prefix: z.string(),
  projectIds: z.array(z.string()),
  defaultProjectId: z.string().nullable(),
  scopes: z.array(z.enum(mcpAPIKeyScopes)),
  expiresAt: nullableDate,
  revokedAt: nullableDate,
  lastUsedAt: nullableDate,
  createdBy: z.string(),
  createdAt: z.string(),
})

export const stateMCPAPIKeyListResponseSchema = z.array(
  stateMCPAPIKeyResponseSchema,
)

export const createStateMCPAPIKeyResponseSchema = z.object({
  key: z.string().min(1),
  apiKey: stateMCPAPIKeyResponseSchema,
})

export type StateMCPAPIKeyResponseSchemaProps = z.infer<
  typeof stateMCPAPIKeyResponseSchema
>
